### PR TITLE
fix: Props are lost when the template replaces the node

### LINKED ISSUE #10636: Type annotation is stripped from Typescript template
## Bug Report

**Current Behavior**
A `@babel/template` template with placeholders and a type annotation generates code without the annotation. A template without placeholders works as expected.

**Input Code**

```js
const template = require('@babel/template').default
const generate = require('@babel/generator').default
const t = require('@babel/types')
​
const withType = template({plugins:['typescript']}).ast`
    const greeting: string = 'Hello';
`
console.log(generate(withType).code)
// => "const greeting: string = 'Hello';"
​
const typeDisappears = template({plugins:['typescript']}).ast`
    const ${t.identifier('greeting')}: string = 'Hello';
`
console.log(generate(typeDisappears).code)
// => "const greeting = 'Hello';"
```

**Expected behavior/code**

I expect both templates above to keep the `: string` type in the generated code.

**Babel Configuration (.babelrc, package.json, cli command)**

No configuration, used from code, see above.

**Environment**
- Babel version(s): 7.6.4
- Node/npm version: Node 12.9.1 / yarn 1.17.3
- Monorepo: no

### PR DESCRIPTION
**Pull request by liuxingbaoyu**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #10636 
| Patch: Bug Fix?          | √
| Major: Breaking Change?  | ?
| Minor: New Feature?      |
| Tests Added + Pass?      | √
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/15286"></a>

### CODE REVIEW

File: packages/babel-template/src/populate.ts [hunk 1]
```diff
@@ -136,10 +149,10 @@ function applyReplacement(
       } else if (Array.isArray(replacement)) {
         items.splice(index, 1, ...replacement);
       } else {
-        items[index] = replacement;
+        assignOrSet(items, index, replacement);
       }
     } else {
-      items[index] = replacement;
+      assignOrSet(items, index, replacement);
```
Comment (liuxingbaoyu) [inline packages/babel-template/src/populate.ts:155]: I didn't find a test example for these two cases.😞

**Review by nicolo-ribaudo — COMMENTED** [review]
I would prefer to explicitly special-case identifiers: if `parent[key].type === "Identifier" && parent[key].typeAnnotation && !replacement.typeAnnotation`, we clone `replacement` and set its `.typeAnnotation` to `parent[key].typeAnnotation`.

**Comment by liuxingbaoyu** [comment]
I'm a little concerned that there will be cases other than the example.

**Comment by liuxingbaoyu** [comment]
Sorry forgot about this! Updated.

**Review by JLHwung — APPROVED** [review]

**Review by nicolo-ribaudo — APPROVED** [review]

**Pull request merged**